source/common/network/io_socket_handle_impl.cc

```diff
@@ -295,5 +278,113 @@ Api::IoCallUint64Result IoSocketHandleImpl::recvmsg(Buffer::RawSlice* slices,
   return sysCallResultToIoCallResult(result);
 }
 
+Api::IoCallUint64Result
+IoSocketHandleImpl::recvmmsg(absl::FixedArray<absl::FixedArray<Buffer::RawSlice>>& slices,
+                             uint32_t self_port, RecvMsgOutput& output) {
+  ASSERT(output.msg_.size() == slices.size());
+  if (slices.size() == 0) {
+    return sysCallResultToIoCallResult(Api::SysCallIntResult{0, EAGAIN});
+  }
+  const uint32_t num_packets_per_mmsg_call = slices.size();
+  absl::FixedArray<mmsghdr> mmsg_hdr(num_packets_per_mmsg_call);
+  absl::FixedArray<absl::FixedArray<struct iovec>> iovs(
+      num_packets_per_mmsg_call, absl::FixedArray<struct iovec>(slices[0].size()));
+  absl::FixedArray<sockaddr_storage> raw_addresses(num_packets_per_mmsg_call);
+  absl::FixedArray<absl::FixedArray<char>> cbufs(num_packets_per_mmsg_call,
+                                                 absl::FixedArray<char>(cmsg_space_));
+
+  auto& os_sys_calls = Api::OsSysCallsSingleton::get();
+  for (uint32_t i = 0; i < num_packets_per_mmsg_call; ++i) {
+    memset(&raw_addresses[i], 0, sizeof(sockaddr_storage));
+    memset(cbufs[i].data(), 0, cbufs[i].size());
+
+    mmsg_hdr[i].msg_len = 0;
+
+    msghdr* hdr = &mmsg_hdr[i].msg_hdr;
+    hdr->msg_name = &raw_addresses[i];
+    hdr->msg_namelen = sizeof(sockaddr_storage);
+    ASSERT(slices[i].size() > 0);
+
+    for (size_t j = 0; j < slices[i].size(); ++j) {
+      iovs[i][j].iov_base = slices[i][j].mem_;
+      iovs[i][j].iov_len = slices[i][j].len_;
+    }
+    hdr->msg_iov = iovs[i].data();
+    hdr->msg_iovlen = slices[i].size();
+    hdr->msg_control = cbufs[i].data();
+    hdr->msg_controllen = cbufs[i].size();
+  }
+
+  const Api::SysCallIntResult result = os_sys_calls.recvmmsg(
+      fd_, mmsg_hdr.data(), num_packets_per_mmsg_call, MSG_TRUNC | MSG_WAITFORONE, nullptr);
+
+  if (result.rc_ <= 0) {
+    return sysCallResultToIoCallResult(result);
+  }
+
+  int num_packets_read = result.rc_;
+
+  for (int i = 0; i < num_packets_read; ++i) {
+    if (mmsg_hdr[i].msg_len == 0) {
+      continue;
+    }
+    RELEASE_ASSERT((mmsg_hdr[i].msg_hdr.msg_flags & MSG_CTRUNC) == 0,
+                   fmt::format("Incorrectly set control message length: {}",
+                               mmsg_hdr[i].msg_hdr.msg_controllen));
+    RELEASE_ASSERT(mmsg_hdr[i].msg_hdr.msg_namelen > 0,
+                   fmt::format("Unable to get remote address from recvmmsg() for fd: {}", fd_));
+    if ((mmsg_hdr[i].msg_hdr.msg_flags & MSG_TRUNC) != 0) {
+      ENVOY_LOG_MISC(warn, "Dropping truncated QUIC packet with size: {}.", mmsg_hdr[i].msg_len);
+      continue;
+    }
+
+    output.msg_[i].msg_len_ = mmsg_hdr[i].msg_len;
+    // Get local and peer addresses for each packet.
+    msghdr& hdr = mmsg_hdr[i].msg_hdr;
+    try {
```

Comment:
A lot of this code is copied, right? Can you share it in a utility function with recvmsg?

Reply:
The fact that we are copying this giant comment would indicate to me that there is some value in a shared utility function. Can we at least move that out as well as the comment?

Reply:
I wrap the try catch block into a helper function getAddressFromSockAddrOrDie()